test/e2e/storage/pv_protection.go

```diff
@@ -0,0 +1,129 @@
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/pkg/util/slice"
+	volumeutil "k8s.io/kubernetes/pkg/volume/util"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/storage/utils"
+)
+
+var _ = utils.SIGDescribe("PV Protection [Feature:StorageProtection]", func() {
+	var (
+		client    clientset.Interface
+		nameSpace string
+		err       error
+		pvc       *v1.PersistentVolumeClaim
+		pv        *v1.PersistentVolume
+		pvConfig  framework.PersistentVolumeConfig
+		pvcConfig framework.PersistentVolumeClaimConfig
+		volLabel  labels.Set
+		selector  *metav1.LabelSelector
+	)
+
+	f := framework.NewDefaultFramework("pv-protection")
+	BeforeEach(func() {
+		client = f.ClientSet
+		nameSpace = f.Namespace.Name
+		framework.ExpectNoError(framework.WaitForAllNodesSchedulable(client, framework.TestContext.NodeSchedulableTimeout))
+
+		// Enforce binding only within test space via selector labels
+		volLabel = labels.Set{framework.VolumeSelectorKey: nameSpace}
+		selector = metav1.SetAsLabelSelector(volLabel)
+
+		pvConfig = framework.PersistentVolumeConfig{
+			NamePrefix: "hostpath-",
+			Labels:     volLabel,
+			PVSource: v1.PersistentVolumeSource{
+				HostPath: &v1.HostPathVolumeSource{
+					Path: "/tmp/data",
+				},
+			},
+		}
+
+		pvcConfig = framework.PersistentVolumeClaimConfig{
+			Annotations: map[string]string{
+				v1.BetaStorageClassAnnotation: "",
+			},
+			Selector: selector,
+		}
+
+		By("Creating a PV")
+		// make the pv definitions
+		pv = framework.MakePersistentVolume(pvConfig)
+		// create the PV
+		pv, err := client.CoreV1().PersistentVolumes().Create(pv)
+		Expect(err).NotTo(HaveOccurred(), "Error creating PV")
+
+		By("Checking that PV Protection finalizer is set")
+		pv, err = client.CoreV1().PersistentVolumes().Get(pv.Name, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred(), "While getting PV status")
+		Expect(slice.ContainsString(pv.ObjectMeta.Finalizers, volumeutil.PVProtectionFinalizer, nil)).To(BeTrue())
+	})
+
+	AfterEach(func() {
+		framework.Logf("AfterEach: Cleaning up test resources.")
+		if errs := framework.PVPVCCleanup(client, nameSpace, pv, pvc); len(errs) > 0 {
+			framework.Failf("AfterEach: Failed to delete PVC and/or PV. Errors: %v", utilerrors.NewAggregate(errs))
+		}
+	})
+
+	It("Verify \"immediate\" deletion of a PV that is not bound to a PVC", func() {
+		By("Deleting the PV")
+		err = client.CoreV1().PersistentVolumes().Delete(pv.Name, metav1.NewDeleteOptions(0))
+		Expect(err).NotTo(HaveOccurred(), "Error deleting PV")
+		framework.WaitForPersistentVolumeDeleted(client, pv.Name, framework.Poll, framework.ClaimProvisionTimeout)
+	})
+
+	It("Verify that PV bound to a PVC is not removed immediately", func() {
+		By("Creating a PVC")
+		pvc := framework.MakePersistentVolumeClaim(pvcConfig, nameSpace)
```

Comment:
We should use `pvc =` instead of `pvc :=` here, otherwise this local variable will override the same name variable in `var` block, and `PVPVCCleanup` method inside `AfterEach` will fail to clean up this pvc object.

Reply:
good catch, thanks. updated